### PR TITLE
311 - limit colleges page and admin pages to admin access

### LINKED ISSUE #311: Security/Bug: non-admin can delete colleges
If a logged in submitter goes to http://localhost:3000/colleges, they can not only see a list of colleges but can edit and destroy colleges.  This should be limited to admin users.

### PR DESCRIPTION
**Pull request by Janell-Huyck**
Fixes #311 

Limits admin pages and "colleges" page to admin-only access.

Previously a non-admin could gain access to the admin pages /colleges, /citations, and to the csv commands.  This PR makes those pages admin-only.

This PR addresses the concern for submitter access to admin pages raised in PR #314.

### What the PR does
Makes the Colleges page/method and Admin pages/methods(csv, citations) restricted to only admin users.

Creates a new "concern", AdminOnlyAccess, which is added to the colleges and admin controller files.  This concern checks to see if a user has an active admin session, and if they don't, they are shown a 404 page.  The check is skipped for the admin login and validate methods because the user would not be logged in as an admin when using them.

### How it is tested
The biggest testing changes center around the colleges controller, which was not previously restricted.  The PR introduces some new "shared examples", `restricts to only admin access`, `restricted access`, and `allowed access` to determine whether or not a user should be able to complete an action (update, create, edit, destroy, etc) based on the user's role (admin, submitter, or none).

The shared examples are called like this:
```
  it_behaves_like 'restricts to only admin access', {
    'index' => :get,
    'show' => :get,
    'new' => :get,
    'edit' => :get,
    'create' => :post,
    'update' => :put,
    'destroy' => :delete
  }
  ```
  The above goes through all the actions for the colleges controller and calls each one of them for each of the roles mentioned.

  Many of the tests inside the colleges_controller_spec.rb file were only looking for a success response upon calling the method.  This functionality is checked inside the shared examples so I removed these duplicate tests from inside the colleges_controller_spec.

The actual logic for the admin-only concern is tested in spec/controllers/concerns/admin_only_access_spec.rb

### Additional changes
- Clarifies (refactors) the "citations" method in the admin controller to make it easier to read and update, and updated the test to work with the find_each syntax used.
- Adds html `status: :unprocessable_entity` to failed college creation or update
- Adds test for csv for 'when an id is sent in the params'
- Adds admin session = true for tests where appropriate
- Minor DRY refactoring for college tests by moving college creation to the top of the file

### What is not covered in this PR
- General submitter-user authorization management (see PR #314 )
- Preventing submitter # 2 from seeing works from submitter # 1 (see PR #323  )

### CODE REVIEW

File: app/controllers/admin_controller.rb [hunk 1]
```diff
@@ -49,15 +55,18 @@ def csv
     end
   end
 
+  # Generates a citations report. This method is restricted to admin
+  # users through the AdminOnlyAccess concern.
   def citations
```
Comment (Janell-Huyck) [inline app/controllers/admin_controller.rb:60]: Refactored the "citations" function in a more Rails-idiomatic fashion.

File: app/controllers/colleges_controller.rb [hunk 2]
```diff
@@ -1,6 +1,8 @@
 # frozen_string_literal: true
 
 class CollegesController < ApplicationController
+  include AdminOnlyAccess
```
Comment (Janell-Huyck) [inline app/controllers/colleges_controller.rb:4]: The Colleges page should have been restricted from regular user access.  Its only purpose is to list, change, and destroy colleges.

File: app/controllers/concerns/admin_only_access.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+# AdminOnlyAccess provides a mechanism to restrict access to certain
+# controller actions to admin users only. It hooks into the controller's
+# lifecycle using a before_action to check the user's admin status.
+module AdminOnlyAccess
```
Comment (Janell-Huyck) [inline app/controllers/concerns/admin_only_access.rb:6]: For any page that requires a user to be an admin in order to access, if the user is not an admin we are redirecting them to a 404 page for security reasons.

File: spec/controllers/admin_controller/admin_controller_citations_spec.rb [hunk 4]
```diff
@@ -6,8 +6,10 @@
   describe 'GET #citations' do
     context 'when admin is logged in' do
       before do
-        allow(College).to receive(:count).and_return(2)
-        allow(controller).to receive(:session).and_return(admin: true)
+        session[:admin] = true
```
Comment (Janell-Huyck) [inline spec/controllers/admin_controller/admin_controller_citations_spec.rb:9]: Refactored to work with the new version that uses Colleges.find_each syntax

File: spec/controllers/admin_controller/admin_controller_citations_spec.rb [hunk 5]
```diff
@@ -21,12 +23,12 @@
...
     context 'when admin is not logged in' do
       before do
-        allow(controller).to receive(:session).and_return(admin: false)
+        session[:admin] = false
       end
 
       it 'redirects to publications_path' do
         get :citations
-        expect(response).to redirect_to(publications_path)
+        expect(response).to have_http_status(:not_found)
```
Comment (Janell-Huyck) [inline spec/controllers/admin_controller/admin_controller_citations_spec.rb:31]: We expect non-admins to be redirected to a 404 page.

File: spec/controllers/admin_controller/admin_controller_csv_spec.rb [hunk 6]
```diff
@@ -34,7 +34,7 @@
     context 'when the user is not an admin' do
       it 'redirects even if a valid format is provided' do
         get(:csv, params: common_params.merge({ format: 'csv' }))
-        expect(response).to redirect_to('/publications')
+        expect(response).to have_http_status(:not_found)
```
Comment (Janell-Huyck) [inline spec/controllers/admin_controller/admin_controller_csv_spec.rb:37]: We expect non-admins to be redirected to a 404 page.

File: spec/controllers/colleges_controller/colleges_controller_create_spec.rb
Comment (Janell-Huyck) [inline spec/controllers/colleges_controller/colleges_controller_create_spec.rb:1]: With the addition of the admin restriction, the test file colleges_controller_spec became too large for Rubocop and had to be split into multiple files.  Each function is given its own file here.

File: spec/controllers/colleges_controller/colleges_controller_destroy_spec.rb
Comment (Janell-Huyck) [inline spec/controllers/colleges_controller/colleges_controller_destroy_spec.rb:1]: With the addition of the admin restriction, the test file colleges_controller_spec became too large for Rubocop and had to be split into multiple files.  Each function is given its own file here.

File: spec/controllers/colleges_controller/colleges_controller_edit_spec.rb
Comment (Janell-Huyck) [inline spec/controllers/colleges_controller/colleges_controller_edit_spec.rb:1]: With the addition of the admin restriction, the test file colleges_controller_spec became too large for Rubocop and had to be split into multiple files.  Each function is given its own file here.

File: spec/controllers/colleges_controller/colleges_controller_index_spec.rb
Comment (Janell-Huyck) [inline spec/controllers/colleges_controller/colleges_controller_index_spec.rb:1]: With the addition of the admin restriction, the test file colleges_controller_spec became too large for Rubocop and had to be split into multiple files.  Each function is given its own file here.

File: spec/controllers/colleges_controller/colleges_controller_new_spec.rb
Comment (Janell-Huyck) [inline spec/controllers/colleges_controller/colleges_controller_new_spec.rb:1]: With the addition of the admin restriction, the test file colleges_controller_spec became too large for Rubocop and had to be split into multiple files.  Each function is given its own file here.

File: spec/controllers/colleges_controller/colleges_controller_show_spec.rb
Comment (Janell-Huyck) [inline spec/controllers/colleges_controller/colleges_controller_show_spec.rb:1]: With the addition of the admin restriction, the test file colleges_controller_spec became too large for Rubocop and had to be split into multiple files.  Each function is given its own file here.

File: spec/controllers/colleges_controller/colleges_controller_update_spec.rb
Comment (Janell-Huyck) [inline spec/controllers/colleges_controller/colleges_controller_update_spec.rb:1]: With the addition of the admin restriction, the test file colleges_controller_spec became too large for Rubocop and had to be split into multiple files.  Each function is given its own file here.

File: spec/controllers/colleges_controller_spec.rb
Comment (Janell-Huyck) [inline spec/controllers/colleges_controller_spec.rb:1]: With the addition of the admin restriction, the test file colleges_controller_spec became too large for Rubocop and had to be split into multiple files.  Each function is given its own file and the original is removed.

**Comment by scherztc** [comment]
@haitzlm 

I ran the test locally.  Do NOT use bundle exec rake spec.  Use bundle exec rspec spec instead.
I tested routes as submitter (/citations, /colleges)
I tested routes as admin (/citations, /colleges, /manage

I did a code read to make sure the concern was use correctly in each of the controllers.
I did a code read on the shared_examples test and other new controller test.

I merged the PR.

**Pull request merged**